proposals/p3762.md

```diff
@@ -0,0 +1,828 @@
+declarations for the same entity may be found. Only one declaration is retained
+for semantic checking and diagnostics. If available, a non-`extern` declaration
+(potentially with a definition) is used; otherwise, the `extern` declaration
+from the earliest import is used.
+
+For a new declaration in the importing file, an imported `extern` declaration is
+always superseded. It is invalid to use the imported declaration before the new
+declaration. It is considered the same entity for type checking, but diagnostics
+should use the local definition.
+
+For a new declaration in the importing file, an imported non-`extern`
+declaration will:
+
+-   If the new declaration is `extern`, the imported non-`extern` is superseded
+    as above. It is invalid to use the imported declaration before the new
+    declaration.
+-   If the new declaration is non-`extern`, the imported declaration must be
+    from the same library.
+    -   The `api` may forward declare when the `impl` defines, or the `impl` may
+        repeat a forward declaration and also define.
+
+### Using `extern` declarations in `extern impl`
+
+An `extern` type may only be used in the type structure of an `impl` if it's an
+`extern impl`. They are incompatible with non-`extern` `impl`s.
+
+Consider two libraries, one defining `A` and declaring `B` as `extern`, and the
+other defining `B` and declaring `A` as `extern`. Neither should be able to
+define an `impl` involving both `A` and `B`, otherwise both could.
+
+### Modifier keywords
+
+When considering various modifiers on a forward declaration versus definition:
+
+-   `extern` is only valid on a forward declaration. Rules are detailed above.
+-   `extend` in `extend impl` is only on the declaration in the class body
+    (whether that is a forward declaration or definition), as described at
+    [Forward `impl` declaration](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#forward-impl-declaration).
+-   Other class, impl, and interface modifiers (`abstract`, `base`, `final`)
+    exist only on the definition, not on the forward declaration.
+-   Function modifiers (`impl`, `virtual`, `default`, `abstract`, `base`,
+    `final`) must match between forward declaration and definition.
+    -   This only affects type-scoped names because they are invalid on
+        namespace names.
+    -   `abstract` won't have a definition so is moot here.
+-   Access modifiers (`private` and `protected`) must match.
+    -   As an exception, an `extern` name may be `private` when the actual name
+        is public.
+        -   This allows a library to forward declare another library's type
+            without allowing clients to depend on its forward declaration.
+        -   On merging, the more public declaration will take precedence, hiding
+            `private extern` declarations.
+    -   This affects both type-scoped names and namespace names.
+
+### Caveats
```

Comment:
```suggestion
### Future work
```
IIRC, the style guide suggests preferring English over Latin loan words.

Reply:
I'd be much more confident that an English reader would understand "quota" and "appendix" than "caveat", and I don't think there is an alternative English word for those.

The reason I was pushing back a little on this is that I would interpret this heading as meaning "Warnings" or "Beware" or maybe "Limitations" or something like that, but that doesn't seem to fit the content here, making me unsure what you mean by this Latin word. To my reading, this isn't a list of warnings, because it's not discussing negative consequences or dangers of this proposal. Rather, this reads to me as a description of some anticipated future work that is not part of this proposal, for which you've considered interactions with this proposal. That's the kind of information I'd expect to see in a "Future work" section.

But in any case, I don't want to block the proposal on this. We've established that proposals are written in the voice of the author, so while I find this heading to be confusing and unhelpful, I won't insist that it's changed, and I approved the proposal on that basis.

Reply:
To me, "future work" is anticipated design work, not just things that people will need to do in the future. "appendix" is a fairly bland name so I've gone with it.